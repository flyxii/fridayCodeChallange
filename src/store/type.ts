export interface State {
  make: string;
  model: string;
}

export type UPDATE_MAKE = { type: Action.UPDATE_MAKE; payload: string };
export type UPDATE_MODEL = { type: Action.UPDATE_MODEL; payload: string };

export enum Action {
  UPDATE_MAKE,
  UPDATE_MODEL,
}

export type Actions = UPDATE_MAKE | UPDATE_MODEL;
