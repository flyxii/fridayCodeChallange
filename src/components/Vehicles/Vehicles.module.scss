@use 'libs/styles/dataWrapper.scss' with (
  $width: 50%,
);
@import "libs/styles/base.scss";

.vehicle{
  @extend %bodyText;
  padding: 16px;
  background-color: $lightGrey;
  margin-bottom: 24px;
}

.vehicleName{
  @extend %H4;
  margin-bottom: 12px;
}