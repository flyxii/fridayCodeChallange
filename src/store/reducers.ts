import { State, Actions, Action } from "./type";

export const reducer = (state: State, action: Actions): State => {
  switch (action.type) {
    case Action.UPDATE_MAKE:
      return { ...state, make: action.payload };
    case Action.UPDATE_MODEL:
      return { ...state, model: action.payload };
    default:
      return state;
  }
};
