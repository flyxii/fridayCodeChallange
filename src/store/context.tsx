/**
 * The useContext hook to create a store that could share data among components
 */

import React, { createContext, useReducer } from "react";
import { reducer } from "./reducers";
import { State, Actions } from "./type";

const initialState: State = {
  make: "",
  model: "",
};

const AppContext = createContext<{
  state: State;
  dispatch: React.Dispatch<Actions>;
}>({
  state: initialState,
  dispatch: () => null,
});

const AppProvider: React.FC = ({ children }) => {
  const [state, dispatch] = useReducer(reducer, initialState);

  return <AppContext.Provider value={{ state, dispatch }}>{children}</AppContext.Provider>;
};

export { AppContext, AppProvider };
