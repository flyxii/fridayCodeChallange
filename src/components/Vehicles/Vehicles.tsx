/**
 * Component to fetch the list of vehicle and display the return data to end user
 * 
 */

import React, { useContext } from "react";
import useSWR from "swr";
import { AppContext } from "store/context";
import Loading from "libs/images/Loading.svg";
import styles from "./Vehicles.module.scss";

type Vehicle = {
  bodyType: string;
  engineCapacity: number;
  enginePowerKW: number;
  enginePowerPS: number;
  fuelType: string;
  make: string;
  model: string;
};

const Vehicles: React.FC = () => {
  /**
   * Connect to the store
   */
  const { state } = useContext(AppContext);
  const { make, model } = state;

  /**
   * Fetch vehicles data from api
   */
  const { data, error } = useSWR<Vehicle[], Error>(`http://localhost:8080/api/vehicles?make=${make}&model=${model}`);

  /**
   * Show empty box if make and model is falsy
   */
  if (!make || !model) return <div className={styles.error} data-testid="empty-wrapper"></div>;

  /**
   * Show loading icon if waiting for data to come or error return from api
   */
  if (error || !data)
    return (
      <div className={styles.error}>
        <img src={Loading} className={styles.loading} alt="Loading" data-testid="loading" />
      </div>
    );

  /**
   * Show error message if no data return from api
   */
  if (data.length === 0) return (
    <div className={styles.error} data-testid="no-data">
      No data found
    </div>
  );

  /**
   * Show the list of vehicles
   */
  return (
    <div className={styles.wrapper}>
      <div className={styles.title}>Vehicles</div>
      <div className={styles.scrollWrapper}>
        {data.map((item, i) => (
          <div className={styles.vehicle} key={i}>
            <div className={styles.vehicleName}>{`${make} ${model}`}</div>
            <div>Body : {item.bodyType}</div>
            <div>Fuel : {item.fuelType}</div>
            <div>Engine Power PS : {item.enginePowerPS}</div>
            <div>Engine Power KW : {item.enginePowerKW}</div>
            <div>Engine Capacity : {item.engineCapacity}</div>
          </div>
        ))}
      </div>
    </div>
  );
};

export default Vehicles;
