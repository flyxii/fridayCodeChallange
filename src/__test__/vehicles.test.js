import * as React from "react";
import { SWRConfig } from "swr";
import fetch from "jest-fetch-mock";
import { render, waitFor } from "@testing-library/react";
import { AppContext } from "store/context";
import Vehicles from "components/Vehicles";
import fetcher from "libs/fetcher";

const AppContextWithoutValue = () => (
  <AppContext.Provider value={{ state: { make: "", model: "" } }}>
    <SWRConfig value={{ dedupingInterval: 0, shouldRetryOnError: false, fetcher: fetcher }}>
      <Vehicles />
    </SWRConfig>
  </AppContext.Provider>
);

const AppContextWithValue = () => (
  <AppContext.Provider value={{ state: { make: "Ford", model: "Fiesta" } }}>
    <SWRConfig value={{ dedupingInterval: 0, shouldRetryOnError: false, fetcher: fetcher }}>
      <Vehicles />
    </SWRConfig>
  </AppContext.Provider>
);

describe("<Vehicles />", () => {
  beforeEach(() => {
    fetch.resetMocks();
    fetch.doMock();
    jest.useFakeTimers();
  });
  afterEach(() => {
    jest.useRealTimers();
  });

  it("renders without crashing", async () => {
    const { container, asFragment } = render(<AppContextWithValue />);
    await waitFor(() => {
      expect(container.firstChild).toBeInTheDocument();
    });

    expect(asFragment()).toMatchSnapshot();
  });

  it("shows empty wrapper if make value is falsy", async () => {
    fetch.mockReject(() => Promise.reject("server is down"));
    const { getByTestId, asFragment } = render(<AppContextWithoutValue />);

    await waitFor(() => {
      const emptyWrapper = getByTestId(/empty-wrapper/i);
      expect(emptyWrapper).toBeInTheDocument();
    });

    expect(asFragment()).toMatchSnapshot();
  });

  it("shows laoding icon if fetch reject", async () => {
    fetch.mockReject(() => Promise.reject("server is down"));
    const { getByTestId, asFragment } = render(<AppContextWithValue />);

    await waitFor(() => {
      const loading = getByTestId(/loading/i);
      expect(loading).toBeInTheDocument();
      expect(fetch).toHaveBeenCalledTimes(1);
      expect(fetch).toHaveBeenCalledWith("http://localhost:8080/api/vehicles?make=Ford&model=Fiesta", undefined);
    });

    expect(asFragment()).toMatchSnapshot();
  });

  it("shows loading icon in loading stage", async () => {
    fetch.mockResponseOnce("");
    const { getByTestId, asFragment } = render(<AppContextWithValue />);

    await waitFor(() => {
      const loading = getByTestId(/loading/i);
      expect(loading).toBeInTheDocument();
    });

    expect(asFragment()).toMatchSnapshot();
  });

  it("shows message if fetch success with empty array", async () => {
    const mockResponse = [];

    fetch.mockResponseOnce(JSON.stringify(mockResponse));
    const { getByTestId, asFragment } = render(<AppContextWithValue />);

    await waitFor(() => {
      const loading = getByTestId(/no-data/i);
      expect(loading).toBeInTheDocument();
    });
    expect(asFragment()).toMatchSnapshot();
  });

  it("shows data if fetch success", async () => {
    const mockResponse = [
      {
        make: "Ford",
        model: "Fiesta",
        enginePowerPS: 54,
        enginePowerKW: 40,
        fuelType: "Diesel",
        bodyType: "Limousine",
        engineCapacity: 1119,
      },
    ];

    fetch.mockResponseOnce(JSON.stringify(mockResponse));
    const { findAllByText, asFragment } = render(<AppContextWithValue />);

    const car = await findAllByText("Ford Fiesta");

    expect(car).toHaveLength(1);
    expect(asFragment()).toMatchSnapshot();
  });
});
