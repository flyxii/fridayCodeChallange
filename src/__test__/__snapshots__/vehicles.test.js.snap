// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Vehicles /> renders without crashing 1`] = `
<DocumentFragment>
  <div
    class="error"
  >
    <img
      alt="Loading"
      class="loading"
      data-testid="loading"
      src="Loading.svg"
    />
  </div>
</DocumentFragment>
`;

exports[`<Vehicles /> shows data if fetch success 1`] = `
<DocumentFragment>
  <div
    class="wrapper"
  >
    <div
      class="title"
    >
      Vehicles
    </div>
    <div
      class="scrollWrapper"
    >
      <div
        class="vehicle"
      >
        <div
          class="vehicleName"
        >
          Ford Fiesta
        </div>
        <div>
          Body : Limousine
        </div>
        <div>
          Fuel : Diesel
        </div>
        <div>
          Engine Power PS : 54
        </div>
        <div>
          Engine Power KW : 40
        </div>
        <div>
          Engine Capacity : 1119
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<Vehicles /> shows empty wrapper if make value is falsy 1`] = `
<DocumentFragment>
  <div
    class="error"
    data-testid="empty-wrapper"
  />
</DocumentFragment>
`;

exports[`<Vehicles /> shows laoding icon if fetch reject 1`] = `
<DocumentFragment>
  <div
    class="error"
  >
    <img
      alt="Loading"
      class="loading"
      data-testid="loading"
      src="Loading.svg"
    />
  </div>
</DocumentFragment>
`;

exports[`<Vehicles /> shows loading icon in loading stage 1`] = `
<DocumentFragment>
  <div
    class="error"
  >
    <img
      alt="Loading"
      class="loading"
      data-testid="loading"
      src="Loading.svg"
    />
  </div>
</DocumentFragment>
`;

exports[`<Vehicles /> shows message if fetch success with empty array 1`] = `
<DocumentFragment>
  <div
    class="error"
    data-testid="no-data"
  >
    No data found
  </div>
</DocumentFragment>
`;
